lib/Core/BuildEngine.cpp

```diff
@@ -1075,10 +1112,104 @@ class BuildEngineImpl {
       cycleList.pop_back();
       stack.pop_back();
     }
-    assert(!cycleList.empty());
 
-    delegate.cycleDetected(cycleList);
-    cancelRemainingTasks();
+    return cycleList;
+  }
+
+  bool breakCycle(const std::vector<Rule*>& cycleList) {
+    TracingInterval i(EngineQueueItemKind::BreakingCycle);
+
+    // Search the cycle for potential means for breaking the cycle. Right now
+    // we use two principle approaches, force a rule to be built (skipping
+    // scanning its dependencies) and supply a previously built result.
+    for (auto ruleIt = cycleList.rbegin(); ruleIt != cycleList.rend(); ruleIt++) {
+      auto& ruleInfo = getRuleInfoForKey((*ruleIt)->key);
+
+      // If this rule is scanning, try to force a rebuild to break the cycle
+      if (ruleInfo.isScanning()) {
+        // Ask the delegate if we should try to resolve this cycle by forcing
+        // a rule to be built?
+        if (!delegate.shouldResolveCycle(cycleList, &ruleInfo.rule, Rule::CycleAction::ForceBuild))
+          return false;
+
+        if (trace) {
+          trace->cycleForceRuleNeedsToRun(&ruleInfo.rule);
+        }
+
+        // Find the rule scan request, if not already deferred, for this rule
+        // and remove it. If the rule scan request was already deferred, we use
+        // the wasForced condition to ignore it later.
+        auto it = findRuleScanRequestForRule(ruleInfosToScan, &ruleInfo);
+        if (it != ruleInfosToScan.end()) {
+          ruleInfosToScan.erase(it);
+        }
+
+        // mark this rule as needs to run (forced)
+        finishScanRequest(ruleInfo, RuleInfo::StateKind::NeedsToRun);
+        ruleInfo.wasForced = true;
+
+        return true;
+      }
+
+      // Check if this rule has a (potentially) valid previous result and if so
+      // try to provide it to the node requesting it to break the cycle.
+      if (ruleInfo.isInProgressWaiting() && ruleInfo.result.builtAt != 0) {
+        auto* taskInfo = ruleInfo.getPendingTaskInfo();
+
+        // find downstream node in the cycle that wants this input
+        auto& nextRuleInfo = getRuleInfoForKey((*std::next(ruleIt))->key);
+
+        // find the corresponding request on this task
+        auto it = findTaskInputRequestForRule(taskInfo->requestedBy, &nextRuleInfo);
+        if (it == taskInfo->requestedBy.end()) {
+          // this rule has not generated an input request yet, so we cannot
+          // provide a value to it
```

Comment:
When does this happen?

Reply:
This situation occurs right now in the unit test the second time through when we are trying to break the cycle, after we have forced the rule to run in the previous attempt.  It was implicitly handled in this manner in my previous code organization, but is more explicit this way.  That said, I will re-read the code path to confirm my understanding of the sequence of events.

Reply:
Ah yes, I see how this happens.  If the node that will ultimately want this input (nextRuleInfo) has not been scanned yet, it won't have an input request.  How the breaking is working right now, we force the node to build, then we supply it the prior value.   There is likely a good bit of room for improvement here.